packages/reducers/src/combineReducers.test.js

```diff
@@ -0,0 +1,110 @@
+			another: { nested: 'object' },
+			stack: (state = []) => state,
+		});
+		expect(Object.keys(reducer(undefined, { type: 'push' }))).toEqual(['stack']);
+	});
+
+	it('maintains referential equality if the reducers it is combining do', () => {
+		const reducer = combineReducers({
+			child1(state = {}) {
+				return state;
+			},
+			child2(state = {}) {
+				return state;
+			},
+			child3(state = {}) {
+				return state;
+			},
+		});
+		const initialState = reducer(undefined, { type: '@@INIT' });
+		expect(reducer(initialState, { type: 'FOO' })).toBe(initialState);
+	});
+
+	it('does not have referential equality if one of the reducers changes something', () => {
+		const reducer = combineReducers({
+			child1(state = {}) {
+				return state;
+			},
+			child2(state = { count: 0 }, action) {
+				switch (action.type) {
+					case 'increment':
+						return { count: state.count + 1 };
+					default:
+						return state;
+				}
+			},
+			child3(state = {}) {
+				return state;
+			},
+		});
+		const initialState = reducer(undefined, { type: '@@INIT' });
+		expect(reducer(initialState, { type: 'increment' })).not.toBe(initialState);
+	});
+
+	it('should return an updated state when additional reducers are passed to combineReducers', () => {
+		const originalCompositeReducer = combineReducers({ foo });
+		const store = createStore(originalCompositeReducer);
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ foo, bar }));
+		store.dispatch(ACTION);
+		const nextState = store.getState();
+		expect(nextState).not.toBe(initialState);
+	});
+
+	it('should return an updated state when reducers passed to combineReducers are changed', () => {
+		const baz = (state = {}) => state;
+		const originalCompositeReducer = combineReducers({ foo, bar });
+		const store = createStore(originalCompositeReducer);
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ baz, bar }));
+		store.dispatch(ACTION);
+		const nextState = store.getState();
+		expect(nextState).not.toBe(initialState);
+	});
+
+	it('should return the same state when reducers passed to combineReducers are not changed', () => {
+		const originalCompositeReducer = combineReducers({ foo, bar });
+		const store = createStore(originalCompositeReducer);
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ foo, bar }));
+		store.dispatch(ACTION);
+		const nextState = store.getState();
+		expect(nextState).toBe(initialState);
+	});
+
+	it('should preserve state when one or more reducers are removed', () => {
+		const originalCompositeReducer = combineReducers({ foo, bar });
+		const store = createStore(originalCompositeReducer);
+		store.dispatch(ACTION);
+		const initialState = store.getState();
+		store.replaceReducer(combineReducers({ bar }));
+		const nextState = store.getState();
+		expect(nextState).not.toBe(initialState);
```

Comment:
wait wait wait, what are you testing here?

"should preserve state when one or more reducers are removed" does not imply that the state has changed, quite the opposite!

the state should not have changed here (referentially) at all!

please change this test to be `expect(nextState).toBe(initialState)`

i think the solution is to leave out the `nextStateForKey !== previousStateForKey` in `hasChanged = hasChanged || nextStateForKey !== previousStateForKey`

that will make it easier for you to refactor into the `reduce` variant, at least you can practice that :P